### PR TITLE
Storages: limit memory usage of data sharing

### LINKED ISSUE #8564: Data sharing may use more memory than expected
img_v3_026b_758a591a-291e-431f-b853-d18c475950fg

img_v3_026b_73732574-1c95-4c8a-ad51-77e1a74395bg

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: close #8564

Problem Summary:
In terms of design, we hope that by scheduling tasks, the shared data will be immediately read and deleted, so under normal circumstances, it will not occupy additional memory.
However, there is a scenario that would break this assumption: If two requests A and B are merged, request A reads faster and request B reads slowly. 

A will share the data with B, but request B may skip directly without reading and deleting these shared data, because it finds its queue is full (https://github.com/pingcap/tiflash/blob/11c056718a90d4fb283890ee0dc842ffa32555f8/dbms/src/Storages/DeltaMerge/ReadThread/MergedTask.cpp#L81-L84), which may result in consuming much more memory than expected.

### What is changed and how it works?
This PR adds limitations on the size of shared data. It includes two aspects:
- Global shared data size limit.
- Limit on the count of shared data held by each DMFileReader.
  - Why not limit the size of shared data held by each DMFileReader?
  - First, it just holds a smart pointer. This pointer can be shared between multiple requests. How many bytes should it take? That needs some considerations.
  - Second, make changes smaller to pick this PR to other LST branches.
- Also, it supports disabling data sharing and doesn't affect read thead.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  - Run a fast tablescanning query and a slow tablescanning query concurrently.
  - In the image below, 
    - on the left is the code of this PR, the limitation of shared data size is 1GB.
    - on the right is the code of master.
  
image

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by Lloyd-Pottiger** [comment]
Can we adjust the execution order of merged tasks to let the slow task run first? Maybe we can simply order by read_columns.

File: dbms/src/Storages/DeltaMerge/tests/gtest_column_sharing_cache.cpp [hunk 1]
```diff
@@ -50,7 +50,7 @@ TEST(ColumnSharingCacheTest, AddAndGet)
     ColumnSharingCache cache;
 
     auto col = createColumn(8);
-    cache.add(1, 8, col);
+    cache.add(1, 8, col, std::numeric_limits<UInt64>::max());
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/tests/gtest_column_sharing_cache.cpp:53]: Add some case about reaching the limit, then get/del

File: dbms/src/Common/TiFlashMetrics.h [hunk 2]
```diff
@@ -582,6 +582,8 @@ static_assert(RAFT_REGION_BIG_WRITE_THRES * 4 < RAFT_REGION_BIG_WRITE_MAX, "Inva
       F(type_ru_exhausted, {"type", "ru_exhausted"}),                                                                               \
       F(type_add_cache_succ, {"type", "add_cache_succ"}),                                                                           \
       F(type_add_cache_stale, {"type", "add_cache_stale"}),                                                                         \
+      F(type_add_cache_count_limit, {"type", "add_cache_count_limit"}),                                                             \
```
Comment (Lloyd-Pottiger) [inline dbms/src/Common/TiFlashMetrics.h:585]:
```suggestion
      F(type_add_cache_reach_count_limit, {"type", "add_cache_count_limit"}),                                                             \
```

maybe `type_add_cache_reach_count_limit` is better

File: dbms/src/Storages/DeltaMerge/File/DMFileBlockInputStream.cpp [hunk 3]
```diff
@@ -63,6 +65,20 @@ DMFileBlockInputStreamPtr DMFileBlockInputStreamBuilder::build(
...
+    if (!enable_read_thread || max_sharing_column_bytes_for_all <= 0)
+    {
+        // Disable data sharing.
+        max_sharing_column_count = 0;
+    }
+    else if (
+        shared_column_data_mem_tracker != nullptr
+        && std::cmp_greater_equal(shared_column_data_mem_tracker->get(), max_sharing_column_bytes_for_all))
+    {
+        // Memory limitation.
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/File/DMFileBlockInputStream.cpp:77]:
```suggestion
        // The memory used reaches the limitation by running queries, disable the data sharing for this DMFile
```

**Review by Lloyd-Pottiger — APPROVED** [review]

**Review by JaySon-Huang — APPROVED** [review]

**Comment by JaySon-Huang** [comment]
Changed to release note to be "Fix a bug that the memory usage rises significantly when there are slow queries"

**Pull request merged**